test/functional/rpc_whitelist.py

```diff
@@ -92,5 +100,79 @@ def run_test(self):
         self.log.info("Strange test 5")
         assert_equal(200, rpccall(self.nodes[0], self.strange_users[4], "getblockcount").status)
 
+        #rpcwhitelistdefaulttest
+        self.test_rpcwhitelist_default_0_without_whitelist()
+        self.test_rpcwhitelist_default_0_specific_whitelist()
+        self.test_whitelistdefault_1_with_specified_permission()
+        self.test_whitelistdefault_1_without_specified_permission()
+
+    def test_rpcwhitelist_default_0_without_whitelist(self):
+        """
+        Tests that with rpcwhitelistdefault=0 and no entries in rpcwhitelist,
+        all authorized users (with rpcauth entries) are allowed to access all RPC methods
+        """
+        self.log.info("Testing rpcwhitelistdefault=0 with no specified permissions")
+        with open(os.path.join(get_datadir_path(self.options.tmpdir, 0), "bitcoin.conf"), 'a', encoding='utf8') as f:
+            f.write("rpcwhitelistdefault=0\n")
+            f.write("rpcauth=user3:50358aa884c841648e0700b073c32b2e$b73e95fff0748cc0b517859d2ca47d9bac1aa78231f3e48fa9222b612bd2083e\n")
```

Comment:
Instead of hardcoding these new `rpcauth` lines, it might be better to expand the existing `users` list.  This would apply to adjacent methods as well.

Reply:
@tdb3  I've pushed an [update](https://github.com/bitcoin/bitcoin/pull/29858/commits/e02af85abaec67d92e976383777a4b0f2caae4e1). Adding the users to the existing list is a better approach. However, some tests are now failing because users are whitelisted in the run_test function, while the new tests are designed to test instances when a user is not whitelisted.

I thought clearing the bitcoin.conf file before writing the new test would solve this issue, but it doesn't. Below is a sample output for the `bitcoin.conf `file in the `test_rpcwhitelist_default_0_without_whitelist` function:

```
rpcwhitelistdefault=0
rpcauth=user1:50358aa884c841............................................................
rpcauth=user2:8650ba41296f6..........................................
rpcauth=user3:50358aa8..................................................

```
Despite having new permissions, users are still whitelisted, and thus this assertion would fail:

`assert_equal(200, rpccall(self.nodes[0], user[0], "getbestblockhash", user[3]).status)
`
Why would this happen? The new tests are still reading permissions from the` run_test `function and not the new `bitcoin.conf`

Reply:
that's because `with open(os.path.join(get_datadir_path(self.options.tmpdir, 0), "bitcoin.conf"), 'a', encoding='utf8') as f` creates a new bitcoin.conf file inside `bitcoin` but outside `node0` directory. the bitcoin.conf file inside `node0` directory where the actual permissions are read from is unchanged. See[ L65](https://github.com/naiyoma/bitcoin/blob/e02af85abaec67d92e976383777a4b0f2caae4e1/test/functional/rpc_whitelist.py#L65) for example.

Reply:
reverted back to  `with open(self.nodes[0].datadir_path / "bitcoin.conf", "a", encoding="utf8") as f:`